# FINITE FIELD ARITHMETIC 
Paling dasar, sebelum group operation (point addition/point doubling)
Mau bikin LUP decomposition (agak kebayang skrg!!)

In [1]:
_.<Z> = GF(2)[]
F.<z> = GF(2^163, modulus=Z^163+Z^7+Z^6+Z^3+1) #163,7,6,3,0

# SQUARING

In [87]:
bil=0
res=0
# s = z
print("0: 1 --> 1")
for i in range(1,F.degree()):
    bil = z^i
    res = bil^2
    print(i,":", bil,"-->", res)

0: 1 --> 1
1 : z --> z^2
2 : z^2 --> z^4
3 : z^3 --> z^6
4 : z^4 --> z^8
5 : z^5 --> z^10
6 : z^6 --> z^12
7 : z^7 --> z^14
8 : z^8 --> z^16
9 : z^9 --> z^18
10 : z^10 --> z^20
11 : z^11 --> z^22
12 : z^12 --> z^24
13 : z^13 --> z^26
14 : z^14 --> z^28
15 : z^15 --> z^30
16 : z^16 --> z^32
17 : z^17 --> z^34
18 : z^18 --> z^36
19 : z^19 --> z^38
20 : z^20 --> z^40
21 : z^21 --> z^42
22 : z^22 --> z^44
23 : z^23 --> z^46
24 : z^24 --> z^48
25 : z^25 --> z^50
26 : z^26 --> z^52
27 : z^27 --> z^54
28 : z^28 --> z^56
29 : z^29 --> z^58
30 : z^30 --> z^60
31 : z^31 --> z^62
32 : z^32 --> z^64
33 : z^33 --> z^66
34 : z^34 --> z^68
35 : z^35 --> z^70
36 : z^36 --> z^72
37 : z^37 --> z^74
38 : z^38 --> z^76
39 : z^39 --> z^78
40 : z^40 --> z^80
41 : z^41 --> z^82
42 : z^42 --> z^84
43 : z^43 --> z^86
44 : z^44 --> z^88
45 : z^45 --> z^90
46 : z^46 --> z^92
47 : z^47 --> z^94
48 : z^48 --> z^96
49 : z^49 --> z^98
50 : z^50 --> z^100
51 : z^51 --> z^102
52 : z^52 --> z^104
53 : z^53 --> z^106
54

In [88]:
# #versi saved in array
arr = []
bil=0
res=0
# s = z
print("0: 1 --> 1")
for i in range(1,F.degree()):
    bil = z^i
    res = bil^2
#     print(i,":", bil,"-->", res)
    y = str(res).replace("z^","").split(' + ') #.replace("z","1")
    arr.append(y)
#     print(y)
arr 

0: 1 --> 1


[['2'],
 ['4'],
 ['6'],
 ['8'],
 ['10'],
 ['12'],
 ['14'],
 ['16'],
 ['18'],
 ['20'],
 ['22'],
 ['24'],
 ['26'],
 ['28'],
 ['30'],
 ['32'],
 ['34'],
 ['36'],
 ['38'],
 ['40'],
 ['42'],
 ['44'],
 ['46'],
 ['48'],
 ['50'],
 ['52'],
 ['54'],
 ['56'],
 ['58'],
 ['60'],
 ['62'],
 ['64'],
 ['66'],
 ['68'],
 ['70'],
 ['72'],
 ['74'],
 ['76'],
 ['78'],
 ['80'],
 ['82'],
 ['84'],
 ['86'],
 ['88'],
 ['90'],
 ['92'],
 ['94'],
 ['96'],
 ['98'],
 ['100'],
 ['102'],
 ['104'],
 ['106'],
 ['108'],
 ['110'],
 ['112'],
 ['114'],
 ['116'],
 ['118'],
 ['120'],
 ['122'],
 ['124'],
 ['126'],
 ['128'],
 ['130'],
 ['132'],
 ['134'],
 ['136'],
 ['138'],
 ['140'],
 ['142'],
 ['144'],
 ['146'],
 ['148'],
 ['150'],
 ['152'],
 ['154'],
 ['156'],
 ['158'],
 ['160'],
 ['162'],
 ['8', '7', '4', 'z'],
 ['10', '9', '6', '3'],
 ['12', '11', '8', '5'],
 ['14', '13', '10', '7'],
 ['16', '15', '12', '9'],
 ['18', '17', '14', '11'],
 ['20', '19', '16', '13'],
 ['22', '21', '18', '15'],
 ['24', '23', '20', '17'],
 ['26', '25

# RESULT

In [89]:
print(1, ",", 1)
for i in range(len(arr)):
    for j in range(len(arr[i])):
#         a.append((arr[i][j]))
        if arr[i][j] == "1":
#             print(i,",",0) #utk extract doang
            print(i+2,",",1)#utk langsung ke matlab (index sudah disesuaikan)
        elif arr[i][j] == "z":
#             print(i, "lel")
#             print(i,",",1)  #z becomes 1, 1 becomes 0, then all naik 1 buat ke matlab? # utk extract doang
            print(i+2,",",2) #utk langsung ke matlab (index sudah disesuaikan)
        else:
            print(i+2,",",int((arr[i][j]))+1)

1 , 1
2 , 3
3 , 5
4 , 7
5 , 9
6 , 11
7 , 13
8 , 15
9 , 17
10 , 19
11 , 21
12 , 23
13 , 25
14 , 27
15 , 29
16 , 31
17 , 33
18 , 35
19 , 37
20 , 39
21 , 41
22 , 43
23 , 45
24 , 47
25 , 49
26 , 51
27 , 53
28 , 55
29 , 57
30 , 59
31 , 61
32 , 63
33 , 65
34 , 67
35 , 69
36 , 71
37 , 73
38 , 75
39 , 77
40 , 79
41 , 81
42 , 83
43 , 85
44 , 87
45 , 89
46 , 91
47 , 93
48 , 95
49 , 97
50 , 99
51 , 101
52 , 103
53 , 105
54 , 107
55 , 109
56 , 111
57 , 113
58 , 115
59 , 117
60 , 119
61 , 121
62 , 123
63 , 125
64 , 127
65 , 129
66 , 131
67 , 133
68 , 135
69 , 137
70 , 139
71 , 141
72 , 143
73 , 145
74 , 147
75 , 149
76 , 151
77 , 153
78 , 155
79 , 157
80 , 159
81 , 161
82 , 163
83 , 9
83 , 8
83 , 5
83 , 2
84 , 11
84 , 10
84 , 7
84 , 4
85 , 13
85 , 12
85 , 9
85 , 6
86 , 15
86 , 14
86 , 11
86 , 8
87 , 17
87 , 16
87 , 13
87 , 10
88 , 19
88 , 18
88 , 15
88 , 12
89 , 21
89 , 20
89 , 17
89 , 14
90 , 23
90 , 22
90 , 19
90 , 16
91 , 25
91 , 24
91 , 21
91 , 18
92 , 27
92 , 26
92 , 23
92 , 20
93 , 29
93 , 28

# VERIFICATION

In [86]:
(z^162)^2

z^161 + z^12 + z^10 + z^5 + z

Last element mapping:

(z^162)^2 --> z^161 + z^12 + z^10 + z^5 + z

di index matlab jadi

163 , 162

163 , 13

163 , 11

163 , 6

163 , 2

bener alhamdulillah

## BAWAH INI KOTRETAN

In [56]:
int(arr[2][0])

6

In [ ]:

# y.replace(" ","")
y = str(res).replace("z^","").split(' + ') #.replace("z","1")

y

In [ ]:
# t = {}
# group = F.list()
# for i in range(len(group)):
#     if '+' not in str(group[i]):
#         t[group[i]] = (group[i])^2
# #         print(i, group[i],"-->", ((group[i])^2))
# t

In [1]:
for i in range(len(group)):
    if '+' not in str(group[i]):
        t[group[i]] = (group[i])^2
#         print(i, group[i],"-->", ((group[i])^2))
t

NameError: name 'group' is not defined

In [ ]:
# def eval_polynomial(poly, val):
#     xs = [ x.strip().replace('^','**') for x in poly.split('+') ]
#     return sum( [eval(n.replace('x', str(val))) for n in xs] )
listval = list(t.values())
listkey = list(t.keys())

In [ ]:
for i in range(len(t)): #gbs, hrs specify dong (?)
# xs = [ x.strip().replace('^','**') for x in poly.split('+') ]
    xs = [ x.strip().strip('z^') for x in str(listval[i]).split('+') ]
    print(xs)